File: string_calculator.py
import re

def add(nums: str) -> int:
    if not nums:
        return 0
    
    # check the string for delimeters, extract regex pattern for those delimeters and then split the nums such that our resultant output is just the numbers to be added
    if nums.startswith("//"):
        delimiters, nums = nums[2:].split("\n", 1)
        delimiters = re.findall(r"\[([^\]]+)\]", delimiters) or [delimiters]
        pattern = "|".join(map(re.escape, delimiters))
        nums = re.split(pattern, nums)
    else:
        nums = re.split("[,\n]", nums)

    # create an array nums_list consisting only the extracted numbers string, and convert the string to int, make sure to ignore the numbers > 1000
    nums_list = [int(n) for n in nums if int(n) <= 1000]

    # check for negative values
    negatives = [int(n) for n in nums if int(n) < 0]
    if negatives:
        raise ValueError(f"negative numbers not allowed: {','.join(map(str, negatives))}")

    return sum(nums_list)

File: stringCalculator.js
function add(nums) {
    if (!nums) return 0;

    let delimiters = [",", "\n"]; // default delimiters: comma and newline
    // check if a custom delimiter is specified
    if (nums.startsWith("//")) {
        let parts = nums.split("\n"); // split the numbers separately
        let customDelimiters = parts[0].slice(2); // extract delimiters
        let delimiterMatches = customDelimiters.match(/\[([^\]]+)\]/g); // find multiple delimiters in square brackets

        // extract multiple delimiters if matches, else single-character delimiter
        if (delimiterMatches) {
            delimiters = delimiterMatches.map(d => d.slice(1, -1));
        } else {
            delimiters = [customDelimiters];
        }

        // remove delimiter header and keep the number part
        nums = parts.slice(1).join("\n");
    }

    // split the numbers using detected delimiters and then convert the string to individual numbers
    let regex = new RegExp(delimiters.map(d => d.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")).join("|"));
    let numList = nums.split(regex).map(Number);

    // check for negative numbers and throw custom error
    let negatives = numList.filter(n => n < 0);
    if (negatives.length) {
        throw new Error(`negative numbers not allowed: ${negatives.join(",")}`);
    }

    // return the sum of the numbers after ignoring numbers > 1000 (converted from string after extracting delimiters)
    return numList.filter(n => n < 1000).reduce((sum, n) => sum + n, 0);
}

module.exports = add;